components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt

```diff
@@ -18,8 +18,23 @@ internal object TabListReducer {
     fun reduce(state: BrowserState, action: TabListAction): BrowserState {
         return when (action) {
             is TabListAction.AddTabAction -> {
+
+                val updatedTabList = if (action.parent != null) {
+                    val parentIndex = state.tabs.indexOfFirst { it.id == action.parent.id }
+                    if (parentIndex == -1) {
+                        throw IllegalArgumentException("The parent does not exist")
+                    }
+
+                    // Add the child tab next to its parent
+                    val child = action.tab.copy(parentId = action.parent.id)
+                    val childIndex = parentIndex + 1
+                    state.tabs.subList(0, childIndex) + child + state.tabs.subList(childIndex, state.tabs.size)
```

Comment:
nit: Another option may be to use `toMutableList()` and `add()` with an index?

Reply:
True, but I'd prefer if we sticked to immutable types in the reducer. A local mutable list wouldn't be a problem, of course, so it's mostly readability. Personally, I find it easier to read to stick to immutable types when we're copying state everywhere, but don't feel strongly :).

Reply:
Updated.